packages/perseus/src/server-item-renderer.tsx

```diff
@@ -438,30 +439,58 @@ export class ServerItemRenderer
 
         const questionRenderer = (
             <AssetContext.Provider value={contextValue}>
-                <Renderer
-                    keypadElement={this.props.keypadElement}
-                    problemNum={this.props.problemNum}
-                    onInteractWithWidget={this.handleInteractWithWidget}
-                    highlightedWidgets={this.state.questionHighlightedWidgets}
-                    apiOptions={apiOptions}
-                    questionCompleted={this.state.questionCompleted}
-                    reviewMode={this.props.reviewMode}
-                    showSolutions={this.props.showSolutions}
-                    ref={(elem) => {
-                        if (elem != null) {
-                            this.questionRenderer = elem;
-                        }
-                    }}
-                    content={this.props.item.question.content}
+                <UserInputManager
                     widgets={this.props.item.question.widgets}
-                    images={this.props.item.question.images}
-                    linterContext={PerseusLinter.pushContextStack(
-                        this.props.linterContext,
-                        "question",
-                    )}
-                    strings={this.context.strings}
-                    {...this.props.dependencies}
-                />
+                    problemNum={this.props.problemNum ?? 0}
+                >
+                    {({
+                        userInput,
+                        handleUserInput,
+                        initializeUserInput,
+                        restoreUserInputFromSerializedState,
+                    }) => {
+                        this.userInput = userInput;
```

Comment:
This is kind of gross. In a perfect world the consumer of Perseus would wrap `ServerItemRenderer` with `UserInputManager` which would give them access to user input without the imperative `getUserInput` API. What would make this easier to do is to combine context/hooks: the user input could be in a context provider and widgets can consumer their user input directly from the context using a hook.

I don't think this is the right time for that though since I've already spent a ton of time refactoring. So my middle ground was to have a single user input stash and to access it via the existing `getUserInput` API; `this.userInput = userInput;` lets me access that state in `getUserInput`.